snuba/utils/streams/consumers/backends/synchronized.py

```diff
@@ -0,0 +1,291 @@
+                    # replace the current subscription reference with the new
+                    # subscription.
+                    current_subscription = subscription
+
+            # TODO: This timeout was a totally arbitrary descision and may or
+            # may not make sense.
+            # TODO: This may need to be protected from EndOfStream errors,
+            # depending on how that configuration has been set.
+            message = self.__commit_log_consumer_backend.poll(0.1)
+            if message is None:
+                continue
+
+            # If this update matches our subscription (whether or not it is
+            # part of the current assignment does not matter -- it may be part
+            # of a later assignment) and was from a remote consumer group that
+            # we are observing, we need to update the offset state. First, we
+            # check the consumer groups reference (because we can do so without
+            # taking the lock):
+            if message.value.consumer_group in self.__remote_consumer_groups:
+                with self.__subscription.get() as subscription:
+                    # If the current subscription has changed while we were
+                    # waiting for a message, we need to discard this message.
+                    # The next loop iteration will update the consumer
+                    # assignment, and we will begin to rebuild state there.
+                    if subscription is not current_subscription:
+                        continue
+
+                    # If the stream for this message matches the subscription
+                    # topics, apply the update to the subscription offsets.
+                    if any((message.stream in topic) for topic in subscription.topics):
+                        subscription.remote_consumer_group_offsets[
+                            message.value.stream
+                        ][message.value.consumer_group] = message.value.offset
+
+    def __update_stream_states(self) -> None:
+        with self.__subscription.get() as subscription:
+            for stream, offset in self.__backend.tell().items():
+                remote_offset = subscription.get_effective_remote_consumer_group_offset(
+                    stream, self.__remote_consumer_groups
+                )
+                if remote_offset is None or offset >= remote_offset:
+                    self.__backend.pause([stream])
+                elif stream not in self.__paused_streams:
+                    # TODO: It should be safe to call resume on a stream
+                    # that is already consuming, but may have some
+                    # performance drawbacks (I vaguely recall that this may
+                    # flush the consume buffer in the Confluent consumer),
+                    # so it may be worth caching which streams are active
+                    # or paused here, or in the KafkaConsumerBackend.
+                    self.__backend.resume([stream])
```

Comment:
This should even be quite a bit simpler than the last implementation of this concept.

What would be ideal is to be able to ensure this change is called after _any_ change of the stream offset. Having the backend provide a callback for offset mutation might be a more better way to manage the side-effects of offset changes.

This will still need to be called directly like it is in `poll` to reflect changes to the remote offsets, but this would reduce the potential for errors if the backend semantics change.

Also, this name is still bad.